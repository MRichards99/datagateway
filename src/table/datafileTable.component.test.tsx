import React from 'react';
import { shallow as enzymeShallow, mount as enzymeMount } from 'enzyme';
import createShallow from '@material-ui/core/test-utils/createShallow';
import createMount from '@material-ui/core/test-utils/createMount';
import DatafileTable from './datafileTable.component';

describe('DatafileTable component', () => {
  let shallow: typeof enzymeShallow;
  let mount: typeof enzymeMount;
  const row = {
    name: 'Test',
    location: '/test',
    size: 1,
    modTime: new Date('2019-06-10 00:00:00'),
  };

  beforeEach(() => {
    shallow = createShallow({});
    mount = createMount();
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('should render correctly', () => {
    const wrapper = shallow(<DatafileTable />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render details panel correctly', () => {
    const wrapper = shallow(<DatafileTable />);

    const detailPanel = wrapper.prop('detailPanel')(row);
    const detailPanelWrapper = shallow(detailPanel);

    expect(detailPanelWrapper).toMatchSnapshot();
  });

  it('should trigger file download when file download action pressed', () => {
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    const wrapper = shallow(<DatafileTable />);

    const actions = wrapper.prop('actions');
    const downloadAction = actions[0];

    downloadAction.onClick({}, row);
    expect(window.alert).toHaveBeenCalledWith(`Downloading ${row.location}`);
  });

  it('should trigger add to cart when add to cart action pressed', () => {
    // TODO: when enzyme supports hooks with shallow test that cart state changes
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    const wrapper = shallow(<DatafileTable />);

    const actions = wrapper.prop('actions');
    const addToCartAction = actions[1](row);

    addToCartAction.onClick({}, row);
    expect(window.alert).toHaveBeenCalledWith(
      `Added ${row.location} to download cart`
    );
  });

  it('should trigger remove from cart when remove from cart action pressed', () => {
    // TODO: when enzyme supports hooks with shallow test that cart state changes
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    const wrapper = shallow(<DatafileTable />);

    const actions = wrapper.prop('actions');
    const addToCartAction = actions[2](row);

    addToCartAction.onClick({}, row);
    expect(window.alert).toHaveBeenCalledWith(
      `Removed ${row.location} from download cart`
    );
  });
});
