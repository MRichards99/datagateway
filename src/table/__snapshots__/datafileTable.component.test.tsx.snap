// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatafileTable component should render correctly 1`] = `
<Table
  actions={
    Array [
      Object {
        "icon": [Function],
        "onClick": [Function],
        "tooltip": "Download file",
      },
      [Function],
      [Function],
    ]
  }
  columns={
    Array [
      Object {
        "field": "name",
        "title": "Name",
      },
      Object {
        "field": "location",
        "title": "Location",
      },
      Object {
        "field": "size",
        "filtering": false,
        "render": [Function],
        "title": "File Size",
        "type": "numeric",
      },
      Object {
        "field": "modTime",
        "title": "Modified time",
        "type": "datetime",
      },
    ]
  }
  data={
    Array [
      Object {
        "location": "/test1",
        "modTime": 2019-06-09T23:00:00.000Z,
        "name": "Test 2-2-1",
        "size": 500,
      },
      Object {
        "location": "/test2",
        "modTime": 2019-06-10T11:00:00.000Z,
        "name": "Test 2-2-2",
        "size": 9499,
      },
    ]
  }
  detailPanel={[Function]}
  title="Datafile table"
/>
`;

exports[`DatafileTable component should render details panel correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Name: 
    </b>
     
    Test
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      File Size: 
    </b>
     
    1 B
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Location: 
    </b>
     
    /test
  </WithStyles(ForwardRef(Typography))>
</div>
`;
